keyboards/annepro2/protocol.c

```diff
@@ -0,0 +1,117 @@
+/*
+ * (c) 2021 by Tomasz bla Fortuna
+ * License: GPLv2
+ *
+ * This file is shared with Shine firmware.
+ */
+
+#include "protocol.h"
+#include "board.h"
+#include "ch.h"
+#include "hal.h"
+
+/* UART communication protocol state */
+protocol_t proto;
+
+void protoInit(protocol_t *proto, void (*callback)(const message_t *)) {
+  proto->previousId = 0;
+  proto->callback = callback;
+  proto->state = FSA_SYNC_1;
+  proto->errors = 0;
+}
+
+/*
+ * A more robust serial protocol which can handle single dropped characters
+ * during transit. At 115200, transmitting the shortest message takes 0.043ms,
+ * at 9600 - 0.52ms.
+ * TODO: ACKing by msgId instead of blind retrying
+ */
+void protoTx(uint8_t cmd, const unsigned char *buf, int payloadSize,
+             int retries) {
+  static uint8_t msgId = 0;
+  chDbgCheck(payloadSize <= MAX_PAYLOAD_SIZE);
+
+  const uint8_t header[5] = {
+      0x7A, 0x1D, cmd, ++msgId, payloadSize,
+  };
+
+  /* We don't implement ACKs, yet some messages should not be lost. */
+  for (int i = 0; i < retries; i++) {
+    sdWrite(&PROTOCOL_SD, header, sizeof(header));
+    if (payloadSize)
+      sdWrite(&PROTOCOL_SD, buf, payloadSize);
+  }
+}
+
+static inline void messageReceived(protocol_t *proto) {
+    if (proto->buffer.msgId != proto->previousId) {
+        /* It's not a resend / duplicate */
+        proto->callback(&proto->buffer);
+        proto->previousId = proto->buffer.msgId;
+    }
+    proto->state = FSA_SYNC_1;
+}
+
+void protoConsume(protocol_t *proto, uint8_t byte) {
+  switch (proto->state) {
+  case FSA_SYNC_1:
+    if (byte == 0x7A) {
+      proto->state = FSA_SYNC_2;
+    } else {
```

Comment:
I believe having 1-byte key-down events is not required and uart is fast enough. Still, that's possible:
```
void protoConsume(uint8_t byte) {
    switch (proto->state) {
        case FSA_SYNC_1:
            if (byte == 0x7A) {
                proto->state = FSA_SYNC_2;
            } else if (byte & 0x80) {
                /* 1-byte keydown event */
                proto->buffer.command     = CMD_LED_KEY_DOWN;
                proto->buffer.payload[0]  = byte; /* col/row encoded */
                proto->buffer.payloadSize = 1;
                proto->callback(&proto->buffer);
                return;
            } else {
                proto->errors++;
            }
            return;
```

Reply:
it is fine to not have a 1 byte keydown. Even at 9600 we can complete the transfer in  < 10ms which is imo good enough